src/nodes/number.cpp

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2025 Team Dissolve and contributors
+
+#include "nodes/number.h"
+#include <string>
+
+Number::Number(const std::variant<int, double> &value) : value_(value) {}
+
+Number::Number(int value) : value_(value) {}
+
+Number::Number(double value) : value_(value) {}
+
+Number &Number::operator=(const Number &other)
+{
+    value_ = other.value_;
+    return *this;
+}
+
+struct AddNumberVisitor
+{
+    template <class T1, class T2> auto operator()(T1 a, T2 b) const -> Number { return {a + b}; }
+};
+Number Number::operator+(const Number &other) const { return std::visit(AddNumberVisitor{}, value_, other.value_); }
+
+struct AddAssignmentNumberVisitor
+{
+    template <class T1, class T2> auto operator()(T1 a, T2 b) const -> Number { return {a += b}; }
```

Comment:
```suggestion
    template <class T1, class T2> auto operator()(T1 &a, T2 b) const -> Number { return {a += b}; }
```

I believe that `a` will need to be a reference to allow for mutation.

Reply:
On further reflection, it does not *need* to be a reference, since the template can be resolved by references, but I still feel that it may be clearer with making the reference explicit.

Also, this is all moot if we can just use lambdas instead of needing the visitor structs